In [11]:
from fastai.text.all import *

In [6]:
df = pd.read_csv('./expenses.csv')
df.head()

description                     category
0     TRADER JOE S #068             Groceries & Food
1    WHOLEFDS SVL#10707             Groceries & Food
2            Amazon.com                       Amazon
3                OPENAI  Subscriptions & Memberships
4  Amazon.com*NH99Y5J73                       Amazon

In [13]:
dls = TextDataLoaders.from_csv(
    path='.',
    csv_fname='expenses.csv',
    text_col='description',
    label_col='category')
dls.show_batch(max_n=3)

In [34]:
train_size = len(dls.train_ds)
valid_size = len(dls.valid_ds)

print(f"Training size: {train_size}")
print(f"Validation size: {valid_size}")

Training size: 134
Validation size: 33


In [24]:
learn = text_classifier_learner(dls, AWD_LSTM, drop_mult=0.5, metrics=accuracy)

In [25]:
learn.fine_tune(3, 1e-2)

In [26]:
learn.show_results()

In [28]:
learn.save('120523')

Path('models/120523.pth')

In [38]:
learn.predict('AMZN Mktp US*GW42262P3')

('Amazon',
 tensor(0),
 tensor([0.2237, 0.0797, 0.0501, 0.0791, 0.0230, 0.0787, 0.0533, 0.0411, 0.0642,
         0.0867, 0.0381, 0.0646, 0.0579, 0.0597]))

# Model Training Results

## 12-05-2023

Training size: 134
Validation size: 33

| epoch | train_loss | valid_loss | accuracy | time  |
|-------|------------|------------|----------|-------|
| 0     | 2.881874   | 2.531036   | 0.333333 | 00:00 |
| 0     | 2.322587   | 2.417079   | 0.454545 | 00:00 |
| 1     | 2.200765   | 2.248553   | 0.484848 | 00:00 |
| 2     | 2.093007   | 2.179276   | 0.484848 | 00:00 |


# References
1. [Text transfer learning](https://docs.fast.ai/tutorial.text.html)
2. [TextDataLoaders.from_csv](https://docs.fast.ai/text.data.html#textdataloaders.from_csv)